### PR TITLE
installation using  Docker 

### PR DESCRIPTION
**Pull request by DavideBaroliUniLu**
In this pull request, I create a docker container based on phusion image:0.9.19. 
The docker-exec is use to launch the  docker machine with X11 enable. For the structure of dockerfile separation I follow the FeNiCs one. 
For any reference to docker, you can see "Containers for portable, productive and performant scientific computing" ([(http://arxiv.org/abs/1608.07573])

### CODE REVIEW

File: dockerfiles/dev-env-base/.bash_history [hunk 1]
```diff
@@ -0,0 +1,2 @@
+ls
```
Comment (mtezzele) [inline dockerfiles/dev-env-base/.bash_history]: Is this file necessary?

File: dockerfiles/base/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,52 @@
+# Builds a base Docker image that all other pygem project images are
+# built off of (FROM). In this image we control the end user
+# experience, namely; creating the user PyGeM, setting up the users
+# environment and the default entry point and cmds.
+#
+# Authors:
```
Comment (mtezzele) [inline dockerfiles/base/Dockerfile]: Please remove the list of authors. We do not follow this style and github easily shows the author of each file.

File: dockerfiles/dev-env-base/Dockerfile [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# Author Davide Baroli <davide.baroli@uni.lu>
```
Comment (mtezzele) [inline dockerfiles/dev-env-base/Dockerfile]: Please remove the list of authors. We do not follow this style and github easily shows the author of each file.

Comment (DavideBaroliUniLu) [inline dockerfiles/dev-env-base/Dockerfile]: For style ok.

File: dockerfiles/dev-env-base/.bash_history (hunk 1 above)
Comment (DavideBaroliUniLu) [inline dockerfiles/dev-env-base/.bash_history]: No it is not necessary.
The only necessary is the Dockerfile in the two subfolder and the docker-exec.

**Comment by DavideBaroliUniLu** [comment]
The dockerfile allows you to install a container with the environment and PyGeM as a virtual machine. So it aim is devote to distribute the code to everyone independently from the operator system of the local machine. It can be also used for installation on cluster.
How to use:
1. build the container
go in base folder
docker build --tag=pygem-base:phusion .
It compiles and creates the container, then you can push for instance on quay or dockerhub. (with a docker push)
go to dev-env
(pay attention on the first line of the file, "FROM " should be in accordance with the  name of the base container )
docker build --tag=pygem-dev-env:phusion .
Again you can push on open pubblic (quay or dockerhub)
Finally to run the image 
I put the docker-exec file  ( it allows to use X11 - so the gui )
./docker-exec pygem-dev-env:phusion  
Then the user can easy use PyGeM without any "requirement" or difficulty in installation.
In addition if you want to shared document/code with the container you can run
docker run -v $pwd:/home/PyGeM/shared -ti pygem-dev-env:phusion

**Comment by mtezzele** [comment]
Ok thank you very much. It is a very useful PR. There is still the author in dockerfiles/base/Dockerfile file. When you are able to fix docker with travis I will merge. If you need help just ask.

**Comment by mtezzele** [comment]
It would be great if you could add the "how to use docker with pygem" in the readme file.

**Comment by DavideBaroliUniLu** [comment]
I am adding the docker build and run in .travis.ylm. 
I am also pushing the image to   
docker.io/pygemdocker/pygem--base:latest
docker.io/pygemdocker/pygem-env-base:latest
How to use docker with pygem with pre-compiled image?
docker run -ti --rm -e DISPLAY=$DISPLAY -v /tmp/.X11-unix:/tmp/.X11-unix docker.io/pygemdocker/pygem-env-base:latest

**Comment by DavideBaroliUniLu** [comment]
The travis compiles fine. But I am still pushing on docker.io the second container.
you can see the check the uploading in [https://cloud.docker.com/pygemdocker/repository/list]

**Comment by mtezzele** [comment]
What is the difference between the base container and the env-base? 

Remember to remove the author in the base file.

In the travis file I think some code is replicated in the case of a linux machine. The docker image has everything we need to test the package right? So we do not have to install everything twice. Maybe an if statement for the osx case will fix it. Otherwise I just did not get docker entirely and I apologize :)

If you add the how to in the readme I will merge (I think the section Dependencies and installation will fit the purpose). Thanks!

**Comment by DavideBaroliUniLu** [comment]
Base container consists of : pulling phusion (light version of ubuntu 16.04) and set the permission to user PyGeM.
Dev-Env depend on Base container and consist of : dependency library + PyGeM.
Can we specify multiple travis file ? One for classical installation based on conda and one for docker ( just pulling from a repository hosting on cloud as in deal.ii or fenics)? I push the dockerfiles since a developer can adapt them to his personal requirement. 
If the user has win/mac/linux and install docker, then thanks to docker potentiality one has a virtual enviroment with all correct dependency and same performance of full installation.   
The only issue is that it is impossible now to test "via travis" the build of docker machine.

**Comment by luca-heltai** [comment]
Actually, it could be possible to simply let travis build the docker machine, use it to do the testing, and then once building the docker and testing have passed, deploy the newly built docker image to docker-hub.

**Comment by mtezzele** [comment]
@DavideBaroliUniLu are you able to implement the suggestion of Luca?

**Comment by DavideBaroliUniLu** [comment]
I am running the deployment with dockerhub at "[https://hub.docker.com/r/pygemdocker/pygem/]".
I add 2 modification to the travis.ylm file: 

- specification of library version 
- substitution of the docker building in travis.ylm by running one hosted on dockerhub.org.

**Comment by mtezzele** [comment]
I fixed 5 minutes ago the compatibility with the new version of python-occ. A rebase should fix it.

**Comment by mtezzele** [comment]
@DavideBaroliUniLu thanks again for this PR! One last thing: please add to the README.md file in the section "Installation" some information about how to deal with docker in order to get pygem.

File: README.md [hunk 4]
```diff
@@ -58,6 +58,33 @@ To uninstall the package you have to rerun the installation and record the insta
...
+Now we will pull the docker.io/pygemdocker/pygem image from our cloud infrastructure:
+```bash
+>  docker pull docker.io/pygemdocker/pygem:latest
+```
+Docker will pull the latest tag of the image pygemdocker/pygem from docker.io. The download is around 3.246 GB. The  image is a great place to start experimenting with PyGeM and includes all dependencies already compiled for you.
+Once the download is complete you can start PyGeM for the first time. Just run:
+```bash
+>  docker run -ti  pygemdocker/pygem:latest
+```
+To facilitate the devoloping, using the text editor,version control and other tools already installed on your computers,
```
Comment (mtezzele) [inline README.md]: *developing and a space after the comma.

**Comment by mtezzele** [comment]
Please perform either a squash or a rebase (better both but I can do one of them via browser). Then I will merge. Thank you!

**Comment by DavideBaroliUniLu** [comment]
I perform both rebase with squash. Now it is pretty better.

**Comment by mtezzele** [comment]
I don't think you did. See [here](https://github.com/mathLab/PyGeM/network). You have to rebase with respect to the upstream/master. Ideally the rebase should has the following output: your branch (the yellow one) has to start from my last commit merged. Something like this should work: `git rebase -i mathlab/master`. See [here](https://github.com/edx/edx-platform/wiki/How-to-Rebase-a-Pull-Request) for more details.

**Comment by DavideBaroliUniLu** [comment]
You have right. I get your point.
I have only clean my branch history. But I have this error in rebase operation with respect to the master.
Could not apply 6640a2923f9e9ba955a74cb26c1d4cce34051a54... write method in the RBFParameters class and tests.
I have also fetch it.
What i notice is : the  log with pick is back with respect to one on matlab/master during the rebase opereation. Do you know how to solve it ?

**Comment by mtezzele** [comment]
The problem is that you have merged this branch with another one that is much older than yours. The rebase now produces conflicts. The ugly but effective way is to close this PR, then you have to open another PR with only one commit of the new files you added/edited in this PR. After this I will links the two PR in order to have some history.
